### PR TITLE
Fast distance solver

### PR DESCRIPTION
**Pull request by vladotomov**
Added a normalization-based distance solver. It's very fast (only computes a gradient field and scales the dofs), but the final values are reasonable only around the zero level set.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3268](https://github.com/mfem/mfem/pull/3268) | @vladotomov | @tzanio | @kmittal2 + @bslazarov | 10/31/22 | 11/24/22 | 11/27/22 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by tzanio** [comment]
There is an error in the CI:

```
diffusion.cpp:339:31: error: no match for ‘operator=’ (operand types are ‘mfem::IterativeSolver::PrintLevel’ and ‘int’)
  339 |       dist_func.print_level = 1;
      |                               ^
In file included from ../../linalg/linalg.hpp:30,
                 from ../../mfem.hpp:43,
                 from diffusion.cpp:74:
../../linalg/solvers.hpp:78:11: note: candidate: ‘mfem::IterativeSolver::PrintLevel& mfem::IterativeSolver::PrintLevel::operator=(const mfem::IterativeSolver::PrintLevel&)’
   78 |    struct PrintLevel
      |           ^~~~~~~~~~
```

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Review by atallah727 — APPROVED** [review]

**Review by bslazarov — APPROVED** [review]
Works great without any errors!

File: miniapps/shifted/distance.cpp [hunk 1]
```diff
@@ -287,8 +317,12 @@ int main(int argc, char *argv[])
       auto ds = new PLapDistanceSolver(p, newton_iter);
       dist_solver = ds;
```
Comment (kmittal2) [inline miniapps/shifted/distance.cpp:318]:
```suggestion
      dist_solver = new PLapDistanceSolver(p, newton_iter);
```

**Comment by kmittal2** [comment]
Update CHANGELOG.

File: miniapps/shifted/dist_solver.hpp [hunk 2]
```diff
@@ -70,19 +70,38 @@ class HeatDistanceSolver : public DistanceSolver
...
+
+   public:
+      NormalizationCoeff(ParGridFunction &u_gf) : u(u_gf) { }
+      virtual double Eval(ElementTransformation &T, const IntegrationPoint &ip);
+   };
+
+public:
+   NormalizationDistanceSolver() { }
+
+   void ComputeScalarDistance(Coefficient& u_coeff, ParGridFunction& dist);
```
Comment (kmittal2) [inline miniapps/shifted/dist_solver.hpp:92]: Might be worth commenting here and in the introduction of the miniapp that this solver is supposed to be a computationally cheaper alternative to pLaplacian and is accurate for distance approximations near the zero level set.

**Review by kmittal2 — APPROVED** [review]
Looks good, modulo some minor comments.

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**